datachannel_js.go

```diff
@@ -0,0 +1,256 @@
+func (d *DataChannel) Negotiated() bool {
+	return d.underlying.Get("negotiated").Bool()
+}
+
+// ID represents the ID for this DataChannel. The value is initially
+// null, which is what will be returned if the ID was not provided at
+// channel creation time. Otherwise, it will return the ID that was either
+// selected by the script or generated. After the ID is set to a non-null
+// value, it will not change.
+func (d *DataChannel) ID() *uint16 {
+	return valueToUint16Pointer(d.underlying.Get("id"))
+}
+
+// NOTE(albrow): Apparently not part of the JavaScript WebRTC API
+// // Priority represents the priority for this DataChannel. The priority is
+// // assigned at channel creation time.
+// Priority PriorityType
+
+// ReadyState represents the state of the DataChannel object.
+func (d *DataChannel) ReadyState() DataChannelState {
+	return newDataChannelState(d.underlying.Get("readyState").String())
+}
+
+// BufferedAmount represents the number of bytes of application data
+// (UTF-8 text and binary data) that have been queued using send(). Even
+// though the data transmission can occur in parallel, the returned value
+// MUST NOT be decreased before the current task yielded back to the event
+// loop to prevent race conditions. The value does not include framing
+// overhead incurred by the protocol, or buffering done by the operating
+// system or network hardware. The value of BufferedAmount slot will only
+// increase with each call to the send() method as long as the ReadyState is
+// open; however, BufferedAmount does not reset to zero once the channel
+// closes.
+func (d *DataChannel) BufferedAmount() uint64 {
+	return uint64(d.underlying.Get("bufferedAmount").Int())
+}
+
+// BufferedAmountLowThreshold represents the threshold at which the
+// bufferedAmount is considered to be low. When the bufferedAmount decreases
+// from above this threshold to equal or below it, the bufferedamountlow
+// event fires. BufferedAmountLowThreshold is initially zero on each new
+// DataChannel, but the application may change its value at any time.
+func (d *DataChannel) BufferedAmountLowThreshold() uint64 {
+	return uint64(d.underlying.Get("bufferedAmountLowThreshold").Int())
+}
+
+// valueToDataChannelMessage converts the given value to a DataChannelMessage.
+// It panics with a JavaScriptTypeConverstionError if this is not possible.
+// val should be obtained from MessageEvent.data where MessageEvent is received
+// via the RTCDataChannel.onmessage callback.
+func valueToDataChannelMessage(val js.Value) DataChannelMessage {
+	// If val is of type string, the conversion is straightforward.
+	if val.Type() == js.TypeString {
+		return DataChannelMessage{
+			IsString: true,
+			Data:     []byte(val.String()),
+		}
+	}
+
+	// For other types, we need to first determine val.constructor.name.
+	constructor := val.Get("constructor")
+	if constructor == js.Null() || constructor == js.Undefined() {
+		panic(JavaScriptTypeConverstionError{
```

Comment:
@backkem I'm not sure how to avoid panicking here. For context, this is called during the `OnMessage` handler function. The API doesn't give us any way to return an error. Any ideas?

Reply:
Maybe we just fall back to a default type instead of throwing an error?

Reply:
Hmm.. the problem here arises when we see a JavaScript type for `Message.Data` that we don't know how to convert to a Go type (typically either `[]byte` or `string`). It's not really possible to have a default type in this case.

We could do something like return an empty message, but IMO it would be better to fail fast than cause potentially confusing issues down the line.

Reply:
I wonder if we need to create this many options. AFAIK the DataChannel (protocol level) spec only allows for 2 types of data: string and binary. In combination with the [binaryType](https://w3c.github.io/webrtc-pc/#dom-datachannel-binarytype) attribute of the WebRTC API the binary data may be interpreted as a "blob" or "arraybuffer". So maybe we don't have to account for a large amount of data types. If we can just cover string and whatever the binary-"blob" datatype is we should be good for now. We can look into the "arraybuffer" when implementing the "binaryType" attribute.

Reply:
Okay check the latest commit and let me know what you think. We should be covering every case in the spec.

Reply:
That conversion step seems a bit involved but I don't if/how to do it differently so we can optimize later. LGTM.